### PR TITLE
Refactor Register and Login to MVVM Pattern

### PR DESCRIPTION
**Pull request by SashaNikitich**

### CODE REVIEW

File: WpfTest/Model/LoginModel.cs [hunk 1]
```diff
@@ -1,6 +1,6 @@
-namespace WpfTest.View;
+namespace WpfTest.Model;
 
-public class LoginViewModel
+public abstract class LoginModel
```
Comment (toor1245) [inline WpfTest/Model/LoginModel.cs]: Why `abstract`?

File: WpfTest/View/LoginView.xaml.cs [hunk 2]
```diff
@@ -12,19 +13,19 @@ namespace WpfTest.View
     /// </summary>
     public partial class LoginView : Window
     {
-        private ApplicationContext db;
-        private readonly LoginValidator validator; 
+        ///private ApplicationContext db;
```
Comment (toor1245) [inline WpfTest/View/LoginView.xaml.cs]: please do not send comments to review, you have to delete them, you have history of your code so you can revert code if needed

Comment (SashaNikitich) [reply WpfTest/View/LoginView.xaml.cs]: Thanks for the advice - it was really helpful!

File: WpfTest/ViewModel/LoginViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                _isLoginSuccessful = value;
+                OnPropertyChanged("IsLoginSuccessful");
+            }
+        }
+
+        public ICommand LoginCommand { get; }
+
+        private void OnLogin(object parameter)
+        {
+            var results = _loginValidator.Validate(this);
```
Comment (toor1245) [inline WpfTest/ViewModel/LoginViewModel.cs]: you should validate not ViewModel but specific Model, so in you case is Login Model like this: The main idea to use LoginModel class as property it in xaml:

```xaml
<TextBox Text="{Binding LoginModel.Login,UpdateSourceTrigger=PropertyChanged, ValidatesOnDataErrors=True}"

<components:BindablePasswordBox  Password="{Binding LoginModel.Password}" BorderThickness="0" Background="Transparent" />
```



```C#

public LoginModel 
{
            get => _loginModel;
            set
            {
                   _loginModel = value
                   OnPropertyChanged("LoginModel");
            }
}

public void OnLogin(object obj)
{
     var results = _loginValidator.Validate(LoginModel);
     // ....
}
```

File: WpfTest/ViewModel/LoginViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            if (results.IsValid)
+            {
+                // Simulating a login check
+                if (Login == "admin" && Password == "admin")
+                {
+                    RedirectToMainWindow();
+                }
+            }
+        }
```
Comment (toor1245) [inline WpfTest/ViewModel/LoginViewModel.cs]: consider case when your model is not valid, what should happen

File: WpfTest/ViewModel/LoginViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+        public string Error
+        {
+            get
+            {
+                var results = _loginValidator.Validate(_loginModel);
+                return results != null && results.Errors.Any() ? string.Join(Environment.NewLine, results.Errors.Select(x => x.ErrorMessage).ToArray()) : string.Empty;
+            }
+        }
```
Comment (toor1245) [inline WpfTest/ViewModel/LoginViewModel.cs]: I didn't understand this code, why you duplicate this logic from 100 line? 

For example in metanit https://metanit.com/sharp/wpf/14.php they just use indexator message and Error as just `NotImplementedException`.

probably in your code you make the same validation twice on the same object, I like this approach with `IDataErrorInfo`, but please consider this functionality more detailed what should be in Error and in indexator.

File: WpfTest/ViewModel/RegisterViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+        public string Error
+        {
+            get
+            {
+                var results = _registerValidator.Validate(_registerModel);
+                return results != null && results.Errors.Any() ? string.Join(Environment.NewLine, results.Errors.Select(x => x.ErrorMessage).ToArray()) : string.Empty;
+            }
+        }
+    }
```
Comment (toor1245) [inline WpfTest/ViewModel/RegisterViewModel.cs]: same here

**Comment by toor1245** [comment]
In general, LGTM

So after resolving above comments I think you can merge.

One additional note, your PR should name what you implement. At the moment, your PR name is `LoginValidator`, it means that in your PR you should only add LoginValidator functionality and integrate it, but here you did refactoring to MMVM, introduced Register Validation.

It is my recommendation to write PR using Google Engineering Practices Documentation:
https://google.github.io/eng-practices/review/developer/

**Pull request merged**